File: code/algo/getoffer/go/4/dp_target_minitem.go
package main

import "math"

func coinChange(nums []int, target int) int {
	dp := make([]int, target+1)
	for i := 1; i <= target; i++ {
		dp[i] = math.MaxInt32
	}
	for _, num := range nums {
		for i := 0; i <= target-num; i++ {
			if dp[i] == math.MaxInt32 {
				continue
			}
			dp[i+num] = min(dp[i+num], dp[i]+1)
		}
	}
	if dp[target] == math.MaxInt32 {
		return -1
	}
	return dp[target]
}

File: code/algo/getoffer/c++/4/dp_target_minitem.cpp
#include <algorithm>
#include <climits>
#include <iostream>
#include <vector>

int min(int a, int b) { return (a < b) ? a : b; }

int coinChange(const std::vector<int> &nums, int target) {
  std::vector<int> dp(target + 1, INT_MAX);
  dp[0] = 0;
  for (int num : nums) {
    for (int i = 0; i <= target - num; ++i) {
      if (dp[i] == INT_MAX) {
        continue;
      }
      dp[i + num] = min(dp[i + num], dp[i] + 1);
    }
  }
  return dp[target] == INT_MAX ? -1 : dp[target];
}

int main() {
  std::vector<int> nums = {1, 2, 5};
  int target = 11;
  std::cout << "Minimum coins needed: " << coinChange(nums, target)
            << std::endl;
  return 0;
}
